contracts/utils/cryptography/draft-ERC7739Utils.sol

```diff
@@ -0,0 +1,288 @@
+        bytes32 separator,
+        bytes32 hashedTypedDataNestedStruct
+    ) internal pure returns (bytes32) {
+        return MessageHashUtils.toTypedDataHash(separator, hashedTypedDataNestedStruct);
+    }
+
+    /**
+     * @dev Computes the nested EIP-712 type hash for the given contents type.
+     *
+     * The `contentsTypeName` is the string name in the app's domain before the parentheses
+     * (e.g. Transfer in `Transfer(address to,uint256 amount)`).
+     *
+     * ```solidity
+     * TypedDataSign({contentsTypeName},bytes1 fields,string name,string version,uint256 chainId,address verifyingContract,bytes32 salt,uint256[] extensions){contentsType}
+     * ```
+     *
+     * Requirements:
+     *  - `contentsType` must be a valid EIP-712 type (see {tryValidateContentsType})
+     */
+    // solhint-disable-next-line func-name-mixedcase
+    function NESTED_TYPED_DATA_TYPEHASH(bytes calldata contentsType) internal pure returns (bytes32) {
+        (bool valid, bytes calldata contentsTypeName) = tryValidateContentsType(contentsType);
+        if (!valid) revert InvalidContentsType();
+        return NESTED_TYPED_DATA_TYPEHASH(contentsType, contentsTypeName);
+    }
+
+    /**
+     * @dev Same as {NESTED_TYPED_DATA_TYPEHASH} but with the `contentsTypeName` already validated.
+     *
+     * See {tryValidateContentsType}.
+     */
+    // solhint-disable-next-line func-name-mixedcase
+    function NESTED_TYPED_DATA_TYPEHASH(
+        bytes calldata contentsType,
+        bytes calldata contentsTypeName
+    ) internal pure returns (bytes32) {
+        return
+            keccak256(
+                abi.encodePacked(
+                    "TypedDataSign(",
+                    contentsTypeName,
+                    "bytes1 fields,",
+                    "string name,",
+                    "string version,",
+                    "uint256 chainId,",
+                    "address verifyingContract,",
+                    "bytes32 salt,",
+                    "uint256[] extensions",
+                    ")",
+                    contentsType
+                )
+            );
+    }
+
+    /**
+     * @dev Try to validate the contents type is a valid EIP-712 type.
+     *
+     * A valid `contentsType` is considered invalid if it's empty or it:
+     *
+     * - Starts with a-z or (
+     * - Contains any of the following bytes: , )\x00
+     */
+    function tryValidateContentsType(
+        bytes calldata contentsType
+    ) internal pure returns (bool valid, bytes calldata contentsTypeName) {
+        uint256 contentsTypeLength = contentsType.length;
+        if (contentsTypeLength == 0) return (false, _emptyCalldataBytes());
+
+        // Does not start with a-z or (
+        bytes1 high = contentsType[0];
+        if ((high > 0x60 && high < 0x7b) || high == 0x28) return (false, _emptyCalldataBytes()); // a-z or (
+
+        // Find the start of the arguments
+        uint256 argsStart = contentsType.indexOf(bytes1("("));
```

Comment:
Here there is an implicit calldata to memory copy. Not sure how bad it is, but maybe we want a `indexOf` that operates on calldata bytes.

Reply:
This could be done in a further optimisation. Not blocking.